sycl/include/CL/sycl/property_list.hpp

```diff
@@ -8,277 +8,96 @@
-
-namespace property {
-
-namespace image {
-class use_host_ptr;
-class use_mutex;
-class context_bound;
-} // namespace image
-
-namespace buffer {
-class use_host_ptr;
-class use_mutex;
-class context_bound;
-} // namespace buffer
-
-namespace queue {
-class enable_profiling;
-class in_order;
-} // namespace queue
-
-namespace detail {
-
-// List of all properties' IDs.
-enum PropKind {
-  // Buffer properties
-  BufferUseHostPtr = 0,
-  BufferContextBound,
-  BufferUseMutex,
-
-  // Image properties
-  ImageUseHostPtr,
-  ImageContextBound,
-  ImageUseMutex,
-
-  // Queue properties
-  QueueEnableProfiling,
-  InOrder,
-
-  PropKindSize
-};
-
-// Base class for all properties. Needed to check that user passed only
-// SYCL's properties to property_list c'tor.
-class PropBase {};
-
-// Second base class, needed for mapping PropKind to class and vice versa.
-template <PropKind PropKindT> class Prop;
+class property_list {
 
-// This class is used in property_list to hold properties.
-template <class T> class PropertyHolder {
-public:
-  PropertyHolder() = default;
+  // Base property wrapper type for type erasure
+  class property_wrapper_base {
+  public:
+    virtual ~property_wrapper_base() = default;
+  };
 
-  PropertyHolder(const PropertyHolder &P) {
-    if (P.isInitialized()) {
-      new (m_Mem) T(P.getProp());
-      m_Initialized = true;
-    }
-  }
+  // Polymorphic wrapper around a concrete property type
+  template <typename T> class property_wrapper : public property_wrapper_base {
+  public:
+    T impl;
 
-  ~PropertyHolder() {
-    if (m_Initialized) {
-      T *MemPtr = reinterpret_cast<T *>(m_Mem);
-      MemPtr->~T();
-    }
-  }
+    property_wrapper(T impl) : impl{std::move(impl)} {}
+  };
 
-  PropertyHolder &operator=(const PropertyHolder &Other) {
-    if (this != &Other) {
-      if (m_Initialized) {
-        T *MemPtr = reinterpret_cast<T *>(m_Mem);
-        MemPtr->~T();
-        m_Initialized = false;
-      }
+  // Map property type indicies against type erased container
+  // Note: Memory fragmentation likely as each property is allocated seperately
+  std::unordered_map<std::type_index, std::shared_ptr<property_wrapper_base>>
+      properties;
 
-      if (Other.m_Initialized) {
-        new (m_Mem) T(Other.getProp());
-        m_Initialized = true;
-      }
-    }
-    return *this;
-  }
+  // Add a property to the map
+  template <typename propertyT> void init_property_list(propertyT &&element) {
+    using this_property = typename std::remove_reference<propertyT>::type;
 
-  void setProp(const T &Rhs) {
-    new (m_Mem) T(Rhs);
-    m_Initialized = true;
-  }
+    std::type_index index = std::type_index(typeid(this_property));
 
-  const T &getProp() const {
-    assert(true == m_Initialized && "Property was not set!");
-    const T *MemPtr = reinterpret_cast<const T *>(m_Mem);
-    return *MemPtr;
+    properties[index] = std::make_shared<property_wrapper<this_property>>(
```

Comment:
Do not like the idea of having ~2 allocations on the heap per property.
What was wrong with previous implementation?

Reply:
The problem with the previous implementation was that it needed to know about each property in order to create the tuple with room for them. This created a dependency on `context`, but now that `context` takes a `property_list` as a parameter in its constructor, we get a circular dependence.

I agree that the allocations are very unfortunate, but I cannot think of a better way to break the cycle. I don't think the allocations will be a huge problem as there are generally few properties, but I'm open for alternatives.

Reply:
@steffenlarsen One can employ a raw pointer and check if its null. Isn't it viable option? Another solution for cross-referencing `shared_ptr`'s is to employ `weak_ptr`.

Reply:
I think - in theory - you could use a statically-sized array with room for raw pointers for each property. This would however require the following:

1. `property_list` must know about all the different property types that exist in order to know how big the array must be. This complicates adding new properties as you would both have to add the new property class and tell `property_list` that it exists.
1. Each property will need a unique index into the array. This further complicates the addition of new properties as you would have to find the highest active index and increment it.
1. Copying a `property_list` would either require the properties on the heap to be shared - introducing the problem of which copy is responsible for freeing the memory - or copy the properties on the heap. This quickly becomes more allocations than the implementation in this PR as it uses `shared_ptr`, so only the unordered map needs a new allocation when copied.
1. The implementation in this PR only allocates if there are properties. Likewise what is suggested would also only allocate if there are properties, but it would use more memory on the stack. This is a minor issue, but empty property lists are very common.

I'm not convinced that it would be a better solution.

Reply:
I still don't get why `shared_ptr` is required inside map and what dependency-cycle you're talking about. I believe `element` is a property.

Reply:
> what dependency-cycle you're talking about.

The reason for this change to the implementation of `property_list` is that `context` constructors now have `property_list` parameters. Due to the existence of the `context_bound` property that uses `context`, if `property_list` was to know about the `context_bound` property we would have a dependency cycle in `context -> property_list -> context_bound -> context`. This is broken by making `property_list` oblivious to the actual properties.

> I still don't get why `shared_ptr` is required inside map

When an `unordered_map` is copied, its content is copied with it. This means that if it contains `unique_ptr` it would attempt to copy these, which is not allowed. In case of a copy of the `property_list` raw pointers would only complicate things as we'd either have to copy the actual properties to a new piece of memory or re-implement `shared_ptr` to correctly free the memory once all references are gone.

> I believe `element` is a property.

`element` is the property, which is why we can make a `shared_ptr` from it, but because it is a template function the size of the element type is only known once the template is resolved. It's likely possible to find the largest of the properties and making the elements of the map that exact size, making sure there's enough room for all properties. However, copying would quickly make memory use worse as the map would likely be larger than if it contained `shared_ptr`. It would however be less fragmented.